File: apps/packages/product-domain/src/workflows/validation.ts
import {
  findWorkflowCatalogAgent,
  findWorkflowCatalogModel,
  workflowAgentSupportsGoals,
  workflowEffortOptions,
  type WorkflowAgentCatalog,
  type WorkflowDefinitionDraft,
  type WorkflowValidationIssue,
} from "./definition";

const INPUT_NAME_PATTERN = /^[A-Za-z][A-Za-z0-9_]*$/u;
const TEMPLATE_TOKEN_PATTERN = /(?<!\{)\{\{([^{}]+)\}\}(?!\})/gu;
const MAX_DEFINITION_ITEMS = 64;

export function validateWorkflowDefinitionDraft(
  draft: WorkflowDefinitionDraft,
  catalog: WorkflowAgentCatalog | null | undefined,
): WorkflowValidationIssue[] {
  const issues: WorkflowValidationIssue[] = [];
  if (!draft.title.trim()) {
    issues.push({ path: "title", message: "Add a workflow title." });
  } else if (draft.title.trim().length > 255) {
    issues.push({ path: "title", message: "Workflow titles must be 255 characters or fewer." });
  }
  if (draft.description.trim().length > 20_000) {
    issues.push({
      path: "description",
      message: "Workflow descriptions must be 20,000 characters or fewer.",
    });
  }
  if (draft.inputs.length > MAX_DEFINITION_ITEMS) {
    issues.push({ path: "inputs", message: "Workflows support at most 64 inputs." });
  }

  const inputNames = new Set<string>();
  draft.inputs.forEach((input, inputIndex) => {
    const path = `inputs.${inputIndex}.name`;
    const name = input.name.trim();
    if (name.length > 64) {
      issues.push({ path, message: "Input names must be 64 characters or fewer." });
    } else if (!INPUT_NAME_PATTERN.test(name)) {
      issues.push({
        path,
        message: "Input names must start with a letter and contain only letters, numbers, or underscores.",
      });
    } else if (inputNames.has(name)) {
      issues.push({ path, message: `Input “${name}” is duplicated.` });
    }
    inputNames.add(name);
  });

  if (draft.stages.length === 0) {
    issues.push({ path: "stages", message: "Add at least one stage." });
  } else if (draft.stages.length > MAX_DEFINITION_ITEMS) {
    issues.push({ path: "stages", message: "Workflows support at most 64 stages." });
  }

  draft.stages.forEach((stage, stageIndex) => {
    const stagePath = `stages.${stageIndex}`;
    const agent = findWorkflowCatalogAgent(catalog, stage.harnessConfig.agentKind);
    if (!agent) {
      issues.push({
        path: `${stagePath}.harnessConfig.agentKind`,
        message: "Choose an agent harness from the current catalog.",
      });
    }

    const modelId = stage.harnessConfig.modelId;
    const model = findWorkflowCatalogModel(catalog, stage.harnessConfig.agentKind, modelId);
    if (modelId && !model) {
      issues.push({
        path: `${stagePath}.harnessConfig.modelId`,
        message: "Choose a model supported by this harness.",
      });
    }

    const effort = stage.harnessConfig.effort;
    if (effort && !modelId) {
      issues.push({
        path: `${stagePath}.harnessConfig.effort`,
        message: "Choose a model before setting effort.",
      });
    } else if (effort && model) {
      const allowed = new Set(
        workflowEffortOptions(catalog, stage.harnessConfig.agentKind, model.id)
          .map((option) => option.value),
      );
      if (!allowed.has(effort)) {
        issues.push({
          path: `${stagePath}.harnessConfig.effort`,
          message: "Choose an effort supported by this exact model.",
        });
      }
    }

    if (stage.steps.length === 0) {
      issues.push({ path: `${stagePath}.steps`, message: "Add at least one prompt step." });
    } else if (stage.steps.length > MAX_DEFINITION_ITEMS) {
      issues.push({
        path: `${stagePath}.steps`,
        message: "A stage supports at most 64 prompt steps.",
      });
    }
    stage.steps.forEach((step, stepIndex) => {
      const stepPath = `${stagePath}.steps.${stepIndex}`;
      if (!step.prompt.trim()) {
        issues.push({ path: `${stepPath}.prompt`, message: "Add a prompt." });
      } else if (step.prompt.length > 100_000) {
        issues.push({
          path: `${stepPath}.prompt`,
          message: "Prompts must be 100,000 characters or fewer.",
        });
      }
      validateInputReferences(step.prompt, inputNames, `${stepPath}.prompt`, issues);
      if (step.goal) {
        if (!step.goal.objective.trim()) {
          issues.push({ path: `${stepPath}.goal.objective`, message: "Add a goal objective." });
        } else if (step.goal.objective.trim().length > 20_000) {
          issues.push({
            path: `${stepPath}.goal.objective`,
            message: "Goal objectives must be 20,000 characters or fewer.",
          });
        }
        validateInputReferences(
          step.goal.objective,
          inputNames,
          `${stepPath}.goal.objective`,
          issues,
        );
        if (!workflowAgentSupportsGoals(catalog, stage.harnessConfig.agentKind)) {
          issues.push({
            path: `${stepPath}.goal`,
            message: "This harness does not support goal-driven steps.",
          });
        }
      }
    });
  });

  return issues;
}

function validateInputReferences(
  value: string,
  inputNames: ReadonlySet<string>,
  path: string,
  issues: WorkflowValidationIssue[],
): void {
  const unmatched = value.replace(TEMPLATE_TOKEN_PATTERN, "");
  if (unmatched.includes("{{") || unmatched.includes("}}")) {
    issues.push({ path, message: "Template expression is malformed." });
  }
  for (const token of value.matchAll(TEMPLATE_TOKEN_PATTERN)) {
    const expression = token[1] ?? "";
    const inputName = expression.startsWith("inputs.") ? expression.slice("inputs.".length) : null;
    if (!inputName || !INPUT_NAME_PATTERN.test(inputName)) {
      issues.push({ path, message: `Unsupported template expression “${expression.trim()}”.` });
      continue;
    }
    if (!inputNames.has(inputName)) {
      issues.push({ path, message: `Template references unknown input “${inputName}”.` });
    }
  }
}

File: server/proliferate/server/workflows/domain/validation.py
"""Pure cross-field validation for workflow definition documents."""

from __future__ import annotations

import re
from copy import deepcopy
from dataclasses import dataclass
from typing import Literal, cast

_INPUT_NAME_PATTERN = re.compile(r"^[A-Za-z][A-Za-z0-9_]*$")
_INPUT_REFERENCE_PATTERN = re.compile(r"(?<!\{)\{\{inputs\.([A-Za-z][A-Za-z0-9_]*)\}\}(?!\})")


@dataclass(frozen=True)
class DefinitionIssue:
    path: str
    message: str
    kind: Literal["invalid_definition", "catalog_selection_unavailable"] = "invalid_definition"


@dataclass(frozen=True)
class ValidatedDefinitionDocument:
    inputs: list[dict[str, object]]
    stages: list[dict[str, object]]


def validate_definition_document(
    *,
    inputs: list[dict[str, object]],
    stages: list[dict[str, object]],
) -> ValidatedDefinitionDocument | DefinitionIssue:
    normalized_inputs = deepcopy(inputs)
    normalized_stages = deepcopy(stages)

    input_names: set[str] = set()
    for index, input_definition in enumerate(normalized_inputs):
        name = str(input_definition["name"])
        if not _INPUT_NAME_PATTERN.fullmatch(name):
            return DefinitionIssue(
                path=f"inputs.{index}.name",
                message=(
                    "Input names must start with a letter and contain only letters, numbers, "
                    "and underscores."
                ),
            )
        if name in input_names:
            return DefinitionIssue(
                path=f"inputs.{index}.name",
                message=f"Input name '{name}' is duplicated.",
            )
        input_names.add(name)

    for stage_index, stage in enumerate(normalized_stages):
        harness = cast(dict[str, object], stage["harnessConfig"])
        model_id = harness.get("modelId")
        effort = harness.get("effort")
        if effort is not None and model_id is None:
            return DefinitionIssue(
                path=f"stages.{stage_index}.harnessConfig.effort",
                message="Choose a specific model before setting reasoning effort.",
            )

        steps = cast(list[dict[str, object]], stage["steps"])
        for step_index, step in enumerate(steps):
            for field_name, text in _step_template_fields(step):
                issue = _validate_template(
                    str(text),
                    input_names=input_names,
                    path=f"stages.{stage_index}.steps.{step_index}.{field_name}",
                )
                if issue is not None:
                    return issue

    return ValidatedDefinitionDocument(inputs=normalized_inputs, stages=normalized_stages)


def _step_template_fields(step: dict[str, object]) -> list[tuple[str, str]]:
    fields = [("prompt", str(step["prompt"]))]
    goal = step.get("goal")
    if isinstance(goal, dict):
        fields.append(("goal.objective", str(goal["objective"])))
    return fields


def _validate_template(
    text: str,
    *,
    input_names: set[str],
    path: str,
) -> DefinitionIssue | None:
    referenced_names = _INPUT_REFERENCE_PATTERN.findall(text)
    remaining = _INPUT_REFERENCE_PATTERN.sub("", text)
    if "{{" in remaining or "}}" in remaining:
        return DefinitionIssue(
            path=path,
            message="Templates may only use the exact form '{{inputs.name}}'.",
        )
    unknown = next((name for name in referenced_names if name not in input_names), None)
    if unknown is not None:
        return DefinitionIssue(
            path=path,
            message=f"Template references unknown input '{unknown}'.",
        )
    return None
